### ISSUE TITLE
[BUG] Fix issues in miri with `-Zmiri-tag-raw-pointers`

### ISSUE THREAD

**Issue by Alexhuszagh** [Alexhuszagh/rust-lexical#76]
## Description

A new flag to check for [stacked borrows](https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md) raises a few issues in miri checks.

## Prerequisites

Here are a few things you should provide to help me understand the issue:

- Rust version: nightly-2022-02-24
- lexical version: 0.8.1
- lexical compilation features used: default

## Test case

Please provide a short, complete (with crate import, etc) test case for
the issue, showing clearly the expected and obtained results.

Example test case:

Run the miri checks with the following (on x86_64 Linux):

```bash
MIRIFLAGS="-Zmiri-tag-raw-pointers" rustup run nightly-2022-02-24-x86_64-unknown-linux-gnu cargo miri test --features=radix --tests --release
```

This produces issues in `lexical-parse-float` and `lexical-write-float`.

## Additional Context

The linked pull request #74 seems to partially address the above issues.

## Status

- <s>lexical-parse-float</s>
- lexical-write-float (affects both with and without the `safe` feature).
- Additional test cases

The minimal test for `lexical-write-float` is `f32_roundtrip_test`.

**Comment by Alexhuszagh** [comment]
The issue with the following lines is that we use a `*const u8` to allow a mutable borrow afterwards, to internally shift the data, knowing they cannot overlap. We know this is valid, however, `-Zmiri-tag-raw-pointers` cannot correctly ensure this. I'm currently unsure how to fix this, if @5225225 knows any solutions, please let me know.

```rust
        // Need to shift digits internally, and write the decimal point.
        // First, move the digits by 1 after leading digits.
        // SAFETY: safe if the buffer is large enough to hold the significant digits.
        let count = digit_count - leading_digits;
        unsafe {
            let src = index_unchecked!(bytes[leading_digits..digit_count]).as_ptr();
            let dst = &mut index_unchecked_mut!(bytes[leading_digits + 1..digit_count + 1]);
            copy_unchecked!(dst, src, count);
        }
```

**Comment by Alexhuszagh** [comment]
# Background

It looks like we should be able to fix this, since we have a check prior that:

```rust
if leading_digits >= digit_count {
    ...
} else {
    // Need to shift digits internally, and write the decimal point.
    // First, move the digits by 1 after leading digits.
    // SAFETY: safe if the buffer is large enough to hold the significant digits.
    let count = digit_count - leading_digits;
    unsafe {
        let buf = &mut index_unchecked_mut!(bytes[leading_digits..digit_count + 1]);
        shift_right_1(buf, count);
    }
}
```

# Unsafe Version

This means the compiler should be able to ensure `count > 0`, which can avoid a trivial start condition. If we use the initial unsafe [implementation](https://godbolt.org/z/ahKv9Y3qv), we get the following assembly:

```rust
pub fn shift_right_ptr(bytes: &mut [u8], count: usize) {
    unsafe {
        let src = bytes.get_unchecked(0..count).as_ptr();
        let dst = bytes.get_unchecked_mut(1..count+1).as_mut_ptr();
        std::ptr::copy(src, dst, count);
    }
}

pub fn shift_right_ptr10(bytes: &mut [u8]) {
    shift_right_ptr(bytes, 10);
}
```

```asm
example::shift_right_ptr:
        mov     rsi, rdi
        add     rdi, 1
        jmp     qword ptr [rip + memmove@GOTPCREL]

example::shift_right_ptr10:
        movzx   eax, word ptr [rdi + 8]
        mov     word ptr [rdi + 9], ax
        mov     rax, qword ptr [rdi]
        mov     qword ptr [rdi + 1], rax
        ret
```

With a for loop without bounds checking (the default, unsafe [implementation](https://godbolt.org/z/fxfs88nMz)), we get:

```rust
pub fn shift_right_unsafe(bytes: &mut [u8], count: usize) {
    for i in (0..count).rev() {
        unsafe {
            *bytes.get_unchecked_mut(i+1) = *bytes.get_unchecked(i);
        }
    }
}

pub fn shift_right_unsafe10(bytes: &mut [u8]) {
    shift_right_unsafe(bytes, 10);
}
```

```asm
example::shift_right_unsafe:
        test    rdx, rdx
        je      .LBB0_2
        push    rax
        mov     rsi, rdi
        add     rdi, 1
        call    qword ptr [rip + memmove@GOTPCREL]
        add     rsp, 8
.LBB0_2:
        ret

example::shift_right_unsafe10:
        movzx   eax, word ptr [rdi + 8]
        mov     word ptr [rdi + 9], ax
        mov     rax, qword ptr [rdi]
        mov     qword ptr [rdi + 1], rax
        ret
```

We can see that these two are identical, as long as `count > 0`.

# Safe Version

Unfortunately, this code makes the safe implementation quite slow. The initial [implementation](https://godbolt.org/z/PfGWEb3Ts) is effectively as follows:

```rust
fn shift_right_slice(bytes: &mut [u8], count: usize) {
    assert!(bytes.len() >= count + 1);
    let src = unsafe { std::slice::from_raw_parts(bytes.as_ptr(), count) };
    let dst = &mut bytes[1..count + 1];
    dst.copy_from_slice(src)
}

pub fn shift_right_slice10(bytes: &mut [u8]) {
    shift_right_slice(bytes, 10);
}
```

```asm
example::shift_right_slice10:
        push    rax
        cmp     rsi, 10
        jbe     .LBB0_1
        movzx   eax, word ptr [rdi + 8]
        mov     word ptr [rdi + 9], ax
        mov     rax, qword ptr [rdi]
        mov     qword ptr [rdi + 1], rax
        pop     rax
        ret
.LBB0_1:
        lea     rdi, [rip + .L__unnamed_1]
        lea     rdx, [rip + .L__unnamed_2]
        mov     esi, 42
        call    qword ptr [rip + core::panicking::panic@GOTPCREL]
        ud2
```

The new [implementation](https://godbolt.org/z/ExEqGozET) is effectively as follows:

```rust
fn shift_right_index(bytes: &mut [u8], count: usize) {
    assert!(bytes.len() >= count + 1);
    for i in (0..count).rev() {
        bytes[i+1] = bytes[i];
    }
}

pub fn shift_right_index10(bytes: &mut [u8]) {
    shift_right_index(bytes, 10);
}
```

```asm
example::shift_right_index10:
        push    rax
        cmp     rsi, 10
        jbe     .LBB0_1
        mov     eax, dword ptr [rdi + 6]
        mov     dword ptr [rdi + 7], eax
        movzx   eax, word ptr [rdi + 4]
        mov     word ptr [rdi + 5], ax
        movzx   eax, word ptr [rdi]
        movzx   ecx, word ptr [rdi + 2]
        mov     word ptr [rdi + 3], cx
        mov     word ptr [rdi + 1], ax
        pop     rax
        ret
.LBB0_1:
        lea     rdi, [rip + .L__unnamed_1]
        lea     rdx, [rip + .L__unnamed_2]
        mov     esi, 42
        call    qword ptr [rip + core::panicking::panic@GOTPCREL]
        ud2
```

This might not be too bad, however, a lot of unnecessary cruft has been added in. In the grand scheme of things, it's probably not too much overhead.

**Comment by Alexhuszagh** [comment]
Fixed as of 3cae06e, which added the relevant tests and patches for the write float functionality.

**Comment by Alexhuszagh** [comment]
Fixed in the released [v0.8.3](https://crates.io/crates/lexical-core/0.8.3).
